### ISSUE TITLE
New OIDC user registrations do not bind email addresses when user consent is required at registration time.

### ISSUE THREAD

**Issue by Half-Shot** [matrix-org/synapse#16317]
This is a bit of an edge case, so please bear with. The parameters are:

- You have enabled `consent.user_consent_at_registration` in your config.
- You have OIDC auth enabled. You allow users to register via OIDC.
- You have set `localpart_template`, `displayname_template` and `email_template`  in the OIDC config so that users can automatically map their emails.

When this is set, the mapped email is always ignored on registration. Specifically you'll get a log line like:

```
2023-09-12 23:08:50,238 - synapse.handlers.sso - 1074 - INFO - GET-219194- [session vsrKfTioDugwCcQw] Registered userid @foo:example.com with attributes UserAttributes(localpart='foo', confirm_localpart=False, display_name='Mr Foo', picture=None, emails=())
```

The OIDC code is quite hard to follow, but I have done my best to understand it. The flow I believe is as follows:

#### /_synapse/client/oidc/callback

- The user logs in, and does the OIDC authentication part. They are sent to `/_synapse/client/oidc/callback`.
- Synapse sees this, works out that the user doesn't exist yet ([here](https://github.com/matrix-org/synapse/blob/ad3f43be9a597dd4fdf59e0a95e4630e7b9502fe/synapse/handlers/sso.py#L482)).
- Because  `user_consent_at_registration` is `true`, the user is directed to `/_synapse/client/new_user_consent` in `_get_url_for_next_new_user_step`.
- `_redirect_to_next_new_user_step` will store the mapped emails in the session object.

#### /_synapse/client/new_user_consent

- The user is asked to consent, and does so.
- Nothing terribly exciting happens to the session state but eventually a new url is generated via `_get_url_for_next_new_user_step`. Because the current stage passes in a `session` object, we are sent to `/_synapse/client/sso_register`.

#### /_synapse/client/sso_register

- This calls `register_sso_user`, which looks at `session.emails_to_use`, rather than `session.emails` that was set in `/_synapse/client/oidc/callback`.  `emails_to_use` is *expected* to be set during `/_synapse/client/pick_username` but because we define a localpart template already, we are never sent to this page.
- The user is created with no bound emails because `emails_to_use` is not set.

### Proposal

We should probably fall back to `session.emails` when `session.emails_to_use` has not been defined.

**Comment by grantm** [comment]
I can't believe the coincidence that you happened to log this issue 6 hours before I started investigating why none of our user accounts got an email address from our Keycloak OIDC server.  It turns out that we have the `user_consent_at_registration` setting enabled in our config.

**Comment by clokep** [comment]
> We should probably fall back to `session.emails` when `session.emails_to_use` has not been defined.

I spoke with @Half-Shot about this separately, but we need to be careful *not* to override people unselecting all their emails *when choosing their username*.
